word_segmentation.py

```diff
@@ -0,0 +1,1252 @@
+    return out_str
+
+
+def get_burmese_text(filename):
+    """
+    This function first combine all lines in a file where each two lines are separated with a space, and then uses ICU
+    to segment the new long string.
+    Note: Because in some of the Burmese texts some lines start with glyphs that are not valid, I first combine all
+    lines and then segment them, rather than first segmenting each line and then combining them. This can result in a
+    more robust segmentation. Eample: see line 457457 of the my_train.txt
+    of the
+    Args:
+        filename: address of the input file
+    """
+    words_break_iterator = BreakIterator.createWordInstance(Locale.getUS())
+    out_str = ""
+    line_counter = 0
+    with open(filename) as f:
+        for line in f:
+            line_counter += 1
+            line = line.strip()
+            if is_ascii(line):
+                continue
+            if len(out_str) == 0:
+                out_str = line
+            else:
+                out_str = out_str + " " + line
+
+    words_break_iterator.setText(out_str)
+    icu_word_brkpoints = [0]
+    for brkpoint in words_break_iterator:
+        icu_word_brkpoints.append(brkpoint)
+    out_str = get_segmented_string(out_str, icu_word_brkpoints)
+    return out_str
+
+
+def get_file_text(filename):
+    """
+    Gives a long string, that contains all lines (separated by a single space) from a file.
+    It removes all texts between pair of tags (<NE>, </NE>) and (<AB>, </AB>), assures that each line starts and ends
+    with a "|", and ignores empty lines, lines with "http" in them, and lines that are all in english (since these are
+    usually not segmented)
+    Args:
+        filename: address of the file
+    """
+    line_counter = 0
+    out_str = ""
+    with open(filename) as f:
+        for line in f:
+            line = clean_line(line)
+            if line == -1:
+                continue
+            if len(out_str) == 0:
+                out_str = line
+            else:
+                out_str = out_str + " " + line
+            line_counter += 1
+    return out_str
+
+
+def get_trainable_data(input_line, graph_clust_ids):
+    """
+    Given a segmented line, extracts x_data (with respect to a dictionary that maps grapheme clusters to integers)
+    and y_data which is a n*4 matrix that represents BIES where n is the length of the unsegmented line. All grapheme
+    clusters not found in the dictionary are set to the largest value of the dictionary plus 1
+    Args:
+        input_line: the unsegmented line
+        graph_clust_ids: a dictionary that stores maps from grapheme clusters to integers
+    """
+    # Finding word breakpoints
+    word_brkpoints = []
+    found_bars = 0
+    for i in range(len(input_line)):
+        if input_line[i] == '|':
+            word_brkpoints.append(i - found_bars)
+            found_bars += 1
+    unsegmented_line = input_line.replace("|", "")
```

Comment:
You do a lot of conversion between the segmented and unsegmented versions of text.  You may want to consider bring this distinction into the type system / make it more object-oriented.  For example, have different classes for segmented and unsegmented strings, with conversion methods to go between them.